roles/openshift_health_checker/openshift_checks/docker_image_availability.py

```diff
@@ -0,0 +1,145 @@
+        },
+        "openshift-enterprise": {
+            "repo": "openshift3",
+            "image": "ose",
+        },
+    }
+
+    def run(self, _, task_vars):
+        is_containerized = get_var(task_vars, "openshift", "common", "is_containerized")
+        openshift_release = get_var(task_vars, "openshift_release")
+        openshift_image_tag = get_var(task_vars, "openshift_image_tag")
+        deployment_type = get_var(task_vars, "deployment_type")
+
+        reg_prefix = self.registry_prefix[deployment_type]
+        required_images = self.get_required_images(is_containerized, reg_prefix, openshift_release)
+
+        # append images with qualified image tags to our list of required images.
+        # these are images with a (v0.0.0.0) tag, rather than a standard release
+        # format tag (v0.0). We want to check this set in both containerized and
+        # non-containerized installations.
+        qualified_imgs = self.qualified_docker_images(self.image_from_reg_prefix(reg_prefix), "v" + openshift_image_tag)
+        required_images.update(qualified_imgs)
+
+        # use docker_image_facts to inspect local images before attempting to use Skopeo
+        required_images = self.inspect_local_images(required_images, task_vars)
+
+        # exit successfully if all required images were present locally
+        if not required_images:
+            return {"changed": False}
+
+        # ensure skopeo image exists or that latest version can be pulled
+        skopeo_update_output = self.update_skopeo_image(task_vars)
+        changed = skopeo_update_output.get("changed", False)
+        if skopeo_update_output.get("failed", False):
+            msg = "Failed to pull required Skopeo image: {msg}".format(msg=skopeo_update_output.get("msg", ""))
+            return {"failed": True, "msg": msg, changed: changed}
+
+        # use docker info to determine which registries
+        # Skopeo should inspect for required images
+        docker_info = self.get_docker_info(task_vars)
+        if not docker_info:
+            return {"failed": True, "msg": "Unable to get docker info. Is `docker-py` installed?", "changed": changed}
+
+        registries = [registry.get("Name", "") for registry in docker_info.get("Registries", {})]
+        if not registries:
+            return {"failed": True, "msg": "Unable to read docker registry data from host.", "changed": changed}
+
+        # inspect all rpm or containerized docker images for every docker registry found
+        while registries and required_images:
+            registry = registries.pop(0)
+            required_images = self.inspect_images(registry, required_images, task_vars)
+
+        # if there are required images left, fail the check
+        if required_images:
+            msg = "Unable to find required images in any docker registries: {imgs}".format(imgs=list(required_images))
+            return {"failed": True, "changed": changed, "msg": msg}
+
+        return {"changed": True}
```

Comment:
We mix so many responsibilities together in `run`:

- query local images
- check is skopeo is available
- get docker daemon information
- prepare a list of registries
- use skopeo to inspect images

Those tasks are in different layers of abstraction.

We could try to write `run` to tell a story, and make it easier to read and understand the intention and what it does on the first sight:

```python
# ...

def run(self, tmp, task_vars):
    required_images = self.required_images(task_vars)
    missing = [name for name in required_images if not self.docker_has_local_image(name)]
    # note: we could perhaps query all images at once, since docker_image_facts takes a list of images
```

Next attempt:

```python
def run(self, tmp, task_vars):
    required_images = self.required_images(task_vars)
    missing_images = set(required_images) - set(self.local_images(required_images, task_vars))

    registries = self.known_docker_registries(task_vars)
    available_images, changed = self.available(missing_images, registries, task_vars)
    unavailable_images = set(missing_images) - set(available_images)

    if unavailable_images:
        return {
            "failed": True,
            "msg": (
                "One or more required images are not available: {}.\n"
                "Configured registries: {}"
            ).format(", ".join(sorted(unavailable_images)), ", ".join(registries)),
            "changed": changed,
        }

    return {"changed": changed}
```

WDYT?